osr2mp4/osrparse/replay.py

```diff
@@ -92,92 +101,92 @@ def bits(n):
 				b = n & (~n+1)
 				yield b
 				n ^= b
-
+	
 		bit_values_gen = bits(self.mod_combination)
 		self.mod_combination = frozenset(Mod(mod_val) for mod_val in bit_values_gen)
 
-	def parse_score_stats(self, replay_data):
-		format_specifier = "<hhhhhhih?i"
-		data = struct.unpack_from(format_specifier, replay_data, self.offset)
-		self.unpack_game_stats(data)
-		self.parse_mod_combination()
-		self.offset += struct.calcsize(format_specifier)
 
-	@staticmethod
-	def __parse_as_int(bytestring):
-		return int.from_bytes(bytestring, byteorder='little')
+	def parse_play_data(self, replay_data: bytes):
+		frames = []
+		lzma_len = self.read_int() # aka self.__replay_length
+		lzma_raw = lzma.decompress(self.read_byte(lzma_len)).decode('ascii')[:-1]
+		events = [event_raw.split('|') for event_raw in lzma_raw.split(',')]
+
+		self.play_data = [
+						ReplayEvent(
+							int(event[0]),
+							float(event[1]),
+							float(event[2]),
+							int(event[3])
+							)
+						for event in events
+						]
+
+
+	### NEW
+	def read_byte(self, length: int = 1):
+		val = self.view[:length]
+		self.view = self.view[length:]
+		return val
+
+	def read_short(self):
+		val = int.from_bytes(self.view[:2], 'little')
+		self.view = self.view[2:]
+		return val
+
+	def read_int(self):
+		val = int.from_bytes(self.view[:4], 'little')
+		self.view = self.view[4:]
+		return val
+
+	def read_float(self):
+		return self.read_int()
+
+	def read_long(self):
+		val = int.from_bytes(self.view[:8], 'little')
+		self.view = self.view[8:]
+		return val
+
+	def read_double(self):
+		return self.read_long()
+
+	def read_uleb128(self):
+		val = shift = 0
 
-	def __decode(self, binarystream):
-		result = 0
-		shift = 0
 		while True:
-			byte = binarystream[self.offset]
-			self.offset += 1
-			result = result |((byte & 0b01111111) << shift)
-			if (byte & 0b10000000) == 0x00:
+			b = int.from_bytes(self.read_byte(), 'little')
+
+
+			val |= ((b & 0b01111111) << shift)
+			if (b & 0b10000000) == 0x00:
 				break
+
 			shift += 7
-		return result
-
-	def parse_player_name(self, replay_data):
-		self.player_name = self.parse_string(replay_data)
-
-	def parse_string(self, replay_data):
-		if replay_data[self.offset] == 0x00:
-			self.offset += Replay.__BYTE
-		elif replay_data[self.offset] == 0x0b:
-			self.offset += Replay.__BYTE
-			string_length = self.__decode(replay_data)
-			offset_end = self.offset + string_length
-			string = replay_data[self.offset:offset_end].decode("utf-8")
-			self.offset = offset_end
-			return string
-		else:
-			#TODO: Replace with custom exception
-			raise Exception("Invalid replay")
 
-	def parse_beatmap_hash(self, replay_data):
-		self.beatmap_hash = self.parse_string(replay_data)
+		return val
 
-	def parse_replay_hash(self, replay_data):
-		self.replay_hash = self.parse_string(replay_data)
+	def read_string(self):
+		if self.read_byte() == 0x00:
+			return ''
```

Comment:
```suggestion
			return ""
```

The existing code seems to use double quotes, I'd recommend trying to stay consistent with that throughout

Reply:
noted